### PR TITLE
Features manager and 4 new features

### PR DESCRIPTION
**Pull request by RaananW**
New feature, **should be discussed before merging**.

Mainly addressing https://github.com/BabylonJS/Babylon.js/issues/7226

Introducing the WebXR features manager. The idea is that XR features are constantly changing so it will be possible to have a few versions of implementation of each feature and let the developer load the one that is working.

It is also possible to use the features manager for existing features such as the teleportation (make teleportation for gamepads, teleportation for AR etc') so it will be easier to add "plugins" or add-ons.

It is currently added to the xr experience helper.

Syntax is simple:

```javascript
// create the XR object:
let xr = await scene.createDefaultXRExperienceAsync({uiOptions: {
        sessionMode: "immersive-ar",
        referenceSpaceType: "local-floor"
    }})

// get the features manager
const fm = xr.baseExperience.featuresManager;

// enable latest hit-test
const xrTest = fm.enableFeature(BABYLON.WebXRHitTestLegacy.Name, "latest", {
    // worldParentNode: xr.baseExperience.container
});
// enable version 1 of plane detection
const xrPlanes = fm.enableFeature(BABYLON.WebXRPlaneDetector.Name, 1, {
    worldParentNode: xr.baseExperience.container
});

// enable latest background remover
const xrBackgroundRemover = fm.enableFeature(BABYLON.WebXRBackgroundRemover.Name);

// just a box we will put in the hit-test result
let b = BABYLON.BoxBuilder.CreateBox('box', { size: 0.1 });
b.rotationQuaternion = new BABYLON.Quaternion();
b.isVisible = false;
b.parent = xr.baseExperience.container;

// on hit test result callback
xrTest.onHitTestResultObservable.add((results) => {
    // show the box
    if (results.length) {
        b.isVisible = true;
        let m = results[0];
        // use the provided (local) transformation matrix and apply it to the box
        m.transformationMatrix.decompose(b.scaling, b.rotationQuaternion, b.position);
        b.position.y += 0.25
    } else {
        b.isVisible = false;
    }
});

// what happens when a new plane is added
xrPlanes.onPlaneAddedObservable.add(plane => {
    let mat;
    if (plane.mesh) {
        mat = plane.mesh.material;
        plane.mesh.dispose(false, false);
    } //else {
    plane.polygonDefinition.push(plane.polygonDefinition[0]);
    plane.mesh = BABYLON.TubeBuilder.CreateTube("tube", { path: plane.polygonDefinition, radius: 0.02, sideOrientation: BABYLON.Mesh.FRONTSIDE, updatable: true }, scene);
    //}
    if (!mat) {
        mat = new BABYLON.StandardMaterial("mat", scene);
        mat.alpha = 0.5;
        mat.diffuseColor = BABYLON.Color3.Random();
    }
    plane.mesh.material = mat;

    plane.mesh.rotationQuaternion = new BABYLON.Quaternion();
    plane.transformationMatrix.decompose(plane.mesh.scaling, plane.mesh.rotationQuaternion, plane.mesh.position);
});
```

Features added:

* Hit Test

Ray-casting to the real world.

Implemented according to the currently-working specs (tested on latest canary for android). Note that a new draft exists and will be version 2.

* Plane Detection

Detect real-world planes in AR.

Implemented according to latest specs. Working on chrome stable for android, chrome canary delivers unexpected results. waiting for chromium to address the bug submitted

* Anchor system

Place anchors in the real world.

Implemented according to implemented undocumented specs. Unstable. Doesn't work on chrome stable, partly working on canary

* Background remover

Removes background elements when entering AR and enabled them when leaving AR.

Functioning stable version that should be used when possible.

### CODE REVIEW

**Comment by deltakosh** [comment]
I love it!! This is REALLY GOOD!

**Comment by deltakosh** [comment]
This will also require an extensive documentation and samples of course :)

File: src/Cameras/XR/features/WebXRAnchorSystem.ts [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+        }
+
+        return <WebXRAnchor>anchor;
+    }
+
+    /**
+     * avoiding using Array.find for global support.
+     * @param xrAnchor the plane to find in the array
+     */
+    private findIndexInAnchorArray(xrAnchor: XRAnchor) {
```
Comment (deltakosh) [inline src/Cameras/XR/features/WebXRAnchorSystem.ts]: private must start with _

File: src/Cameras/XR/features/WebXRBackgroundRemover.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+import { WebXRSessionManager } from '../webXRSessionManager';
+import { AbstractMesh } from '../../../Meshes/abstractMesh';
+import { Observable } from '../../../Misc/observable';
+
+const Name = "xr-background-remover";
+
+/**
+ * Options interface for the background remover plugin
+ */
+export interface WebXRBackgroundRemoverOptions {
```
Comment (deltakosh) [inline src/Cameras/XR/features/WebXRBackgroundRemover.ts]: Interface should start with I

File: src/Cameras/XR/features/WebXRAnchorSystem.ts (hunk 1 above)
Comment (RaananW) [reply src/Cameras/XR/features/WebXRAnchorSystem.ts]: I will go over everything, i think i missed a few of those

**Comment by RaananW** [comment]
Documentation to follow, and if I have the time I think it will make sense to move all XR features to use it (teleportation, gaze, etc'). Maybe not in 4.1?

**Comment by deltakosh** [comment]
Agree, if you have time before 4.1 it will help with back compat (as webXR is still new to 4.1)

**Pull request merged**